### PR TITLE
new /learn page with Code Studio carousel

### PR DESCRIPTION
**Pull request by tanyaparker**
What this PR entails:
- 50/50 mc and star wars on code.org homepage, /learn, teacher dashboard, and studio.code.org
- code studio carousel to show our 8 tutorials, at the top of /learn (also applied to csedweek.org/learn and partner sites like uk.code.org/learn)
- in i18n and mobile, star wars course only goes to /starwarsblock
- third party tutorial changes: moved JavaScript category to top and removed Make Apps category
- created a teacher resources page for Infinity Play Lab and Artist (I made these with Melissa J before so I know what needs to be in them, but I'll verify w/ her on Monday)

Notes:
- The course block needs to be at least ~290px to not look too skinny. So for screen sizes less than 400px, since the carousel control arrows don't fit I put them below the carousel.
- Removed all logic for page_mode "minecraft" or "starwars" and updated to "feature-mc" and "feature-starwars"

Todo:
- [x] swipe touch on mobile
- [x] animate hero & banner to switch between mc & sw
- [x] weird responsiveness between 990px - 1000px
- [x] remove code.org tutorials from 3rd party section (spreadsheet)
- [x] delete starwars_banner_mobile.jpg
- [x] test on mobile, i18n, all browsers (esp IE8), csedweek.org/learn, uk.code.org/learn
- [x] sync pegasus strings

image

image

### CODE REVIEW

**Comment by tanyaparker** [comment]
Oh yeah and I'll also do some cleanup with some leftover in-line css and stuff, so primarily focus on the big picture but please do note nits too if you like.

File: dashboard/app/views/api/hoc_courses.html.haml [hunk 1]
```diff
@@ -1,25 +1,20 @@
 - page_mode = PageMode.get(request)
+- if page_mode == "feature-mc"
+  - first_tutorial = Script::MINECRAFT_NAME
+  - second_tutorial = Script::STARWARS_NAME
+- elsif page_mode == "feature-starwars"
```
Comment (caleybrock) [inline dashboard/app/views/api/hoc_courses.html.haml]: is it possible that when you merge this, page mode won't be "feature-starwars" or "feature-minecraft"? would it be better to have a true else case?

**Comment by caleybrock** [comment]
for the screenshots/design: 
- will the arrows on mobile always appear above the fold? maybe we should keep them above?
- it looks like in the code you got rid of the bee? and if you kept it, it bee now overlaps the tabs.

File: pegasus/sites.v3/code.org/public/index.haml [hunk 2]
```diff
@@ -52,8 +52,10 @@ social:
 - # The homepage gsheet may specify "en" or "not en" for many row types.
 - languageFilter = request.language == "en" ? "en" : "not en"
 
+- page_mode = PageMode.get(request)
```
Comment (caleybrock) [inline pegasus/sites.v3/code.org/public/index.haml:55]: why did you add this some places and remove it other?

File: pegasus/sites.v3/code.org/public/mc.haml [hunk 3]
```diff
@@ -22,34 +22,31 @@ social:
 -# Because of partner requirements, all urls need to use mc, but our internal shortcut is "mchoc".
 -# Hoc.com has a mc.redirect to his page and the router has a check for "code.org/MC" and "hoc.com/MC"
 
-- page_mode = PageMode.get(request)
-
```
Comment (caleybrock) [inline pegasus/sites.v3/code.org/public/mc.haml:26]: don't we need this for it to work?

File: pegasus/sites.v3/code.org/views/homepage_hero.haml [hunk 4]
```diff
@@ -1,12 +1,12 @@
 #hero
-  - page_mode = PageMode.get(request)
-
   -# temporary for hoc2015: using id_s is a quick hack to only pull the hero desired.
   -# Generate a random set of hero images alternating between non-celeb and celeb.
   - heroes = DB[:cdo_homepage].where(id_s: "hero_#{page_mode}", type_s: "widehero", active_s: ["show","hide"], site_s: [request.site,"both"], language_supported_s: ["all", languageFilter]).all
 
-  - if page_mode == "minecraft"
+  - if page_mode == "feature-mc"
```
Comment (caleybrock) [inline pegasus/sites.v3/code.org/views/homepage_hero.haml]: does hadi want these to flip/2 fade in like the learn page? Is that the end goal?

File: pegasus/sites.v3/code.org/views/learn_carousel_block.haml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    .courseblock-tall
+      %a{href: "http://studio.code.org/s/#{id}/reset"}
+        .imgspan
+          %img{src: "/shared/images/courses/logo_tall_#{id}.jpg", style: "width: 100%;"}
+        .clear
+        .heading
+          %h4{style: "font-size: 18px;"}= I18n.t(:"#{id}_name")
+        .clear
+        .smalltext
+          = I18n.t(:"#{id}_shortdescription")
```
Comment (caleybrock) [inline pegasus/sites.v3/code.org/views/learn_carousel_block.haml:20]: did you test this when the descriptions are long/for international?

File: pegasus/sites.v3/code.org/views/learn_carousels.haml [hunk 6]
```diff
@@ -1,19 +1,10 @@
-- page_mode = PageMode.get(request)
-
-- categories = [{tag:"Beginner", name: I18n.t(:carousel_heading_beginners)}, {tag:"JavaScript", name: I18n.t(:carousel_heading_javascript)}, {tag:"Unplugged", name: I18n.t(:carousel_heading_unplugged)}, {tag:"Apps", name: I18n.t(:carousel_heading_devices)}, {tag:"OtherLang", name: I18n.t(:carousel_heading_languages)}, {tag:"MakeApps", name: I18n.t(:carousel_heading_apps_games)}, {tag: "Univ", name: I18n.t(:carousel_heading_university)}, {tag: "Web", name: I18n.t(:carousel_heading_webpages)}, {tag: "Robot", name: I18n.t(:carousel_heading_robots)}, {tag: "IDE", name: I18n.t(:carousel_heading_ide)}]
+- categories = [{tag:"JavaScript", name: I18n.t(:carousel_heading_javascript)}, {tag:"Beginner", name: I18n.t(:carousel_heading_beginners)}, {tag:"Unplugged", name: I18n.t(:carousel_heading_unplugged)}, {tag:"Apps", name: I18n.t(:carousel_heading_devices)}, {tag:"OtherLang", name: I18n.t(:carousel_heading_languages)}, {tag: "Univ", name: I18n.t(:carousel_heading_university)}, {tag: "Web", name: I18n.t(:carousel_heading_webpages)}, {tag: "Robot", name: I18n.t(:carousel_heading_robots)}, {tag: "IDE", name: I18n.t(:carousel_heading_ide)}]
```
Comment (caleybrock) [inline pegasus/sites.v3/code.org/views/learn_carousels.haml:1]: why do you still have beginner here? aren't those just code.org tutorials?

**Comment by caleybrock** [comment]
this image: /images/starwars_banner_mobile.jpg is sliced weirdly. @breville do you have the psd version somewhere? 

I'd say take a look at the test failures, but generally looking good!

File: lib/dynamic_config/page_mode.rb [hunk 7]
```diff
@@ -16,9 +12,10 @@ class PageModeBase
   # @return {string}
   def get(request)
     # If a session page mode is set the 'pm' cookie, return that,
```
Comment (breville) [inline lib/dynamic_config/page_mode.rb:15]: in*

File: lib/dynamic_config/page_mode.rb [hunk 8]
```diff
@@ -16,9 +12,10 @@ class PageModeBase
...
-    # otherwise return the default page mode from DCDO.
-    (request && request.cookies[PAGE_MODE_KEY]) ||
-        DCDO.get('page_mode', DEFAULT_PAGE_MODE)
+    # otherwise return a sw or mc page mode
+    page_mode = rand(2) == 0 ? "feature-starwars" : "feature-mc"
+
+    (request && request.cookies[PAGE_MODE_KEY]) || page_mode
```
Comment (breville) [inline lib/dynamic_config/page_mode.rb]: what if the cookie has a bad value for some reason?  should we check against a whitelist, or like Caley suggests, just always have a fallback to one of the values?

File: pegasus/helpers/page_helpers.rb [hunk 9]
```diff
@@ -29,3 +29,7 @@ def partner_site?()
   partner_sites = CDO.partners.map{|x|x + '.code.org'}
   return partner_sites.include?(request.site)
 end
+
+def mobile?
+  return !(request.user_agent =~ /Mobile/).nil?
```
Comment (breville) [inline pegasus/helpers/page_helpers.rb]: we have a very different check for mobile devices in learn-carousels.haml.  which one do we trust more?

File: pegasus/sites.v3/code.org/public/css/learn-carousel.css [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+.carousel-control.right {
+  right: -25px;
+}
+
+.multi-item-carousel .carousel-inner > .item {
+  transition: 0.6s ease-in-out all;
+}
+
+@media all and (transform-3d), (-webkit-transform-3d) {
```
Comment (breville) [inline pegasus/sites.v3/code.org/public/css/learn-carousel.css:59]: are we okay on back compat all the way back to IE8?

File: pegasus/sites.v3/code.org/public/hourofcode/artist.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
+---
+title: Teacher Resources - Artist
+nav: hoc_teacher_resources_nav
+---
+
+# Artist: An Hour of Code Tutorial
+
+<center><iframe width="600" height="337" src="https://www.youtube.com/embed/MUavAHn-QUg" frameborder="0" allowfullscreen></iframe></center>
```
Comment (breville) [inline pegasus/sites.v3/code.org/public/hourofcode/artist.md:8]: can we use our new video player partial?

File: pegasus/sites.v3/code.org/public/hourofcode/artist.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+## Overview
+
+Draw cool pictures and designs with the Artist!
+
+**Recommended Age Range:** 4-104
+
+**Technology Requirements:** Modern browsers + tablets
+
+<a href="http://studio.code.org/s/artist/reset"><button>Try the tutorial</button></a>
```
Comment (breville) [inline pegasus/sites.v3/code.org/public/hourofcode/artist.md:37]: always good to use CDO functions so that it points at the current environment's version of this link.

File: pegasus/sites.v3/code.org/public/hourofcode/artist.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+<center><iframe width="600" height="337" src="https://www.youtube.com/embed/MUavAHn-QUg" frameborder="0" allowfullscreen></iframe></center>
+
+[col-50]
+
+## Teacher Resources
+
+### Extension Activities:
+
+- [Beyond an Hour of Code](http://code.org/learn/beyond)
```
Comment (breville) [inline pegasus/sites.v3/code.org/public/hourofcode/artist.md:16]: Using /learn/beyond will point to current environment.

File: pegasus/sites.v3/code.org/views/homepage_hero.haml [hunk 14]
```diff
@@ -1,12 +1,12 @@
...
     - heroes = [{:id=>2, :id_s=>"hero_minecraft", :type_s=>"widehero", :param1_s=>"15% 50%", :param2_s=>"celeba", :param3_s=>"bottom", :image_t=>"/images/homepage/minecraft.jpg", :site_s=>"both", :active_s=>"show", :language_supported_s=>"all"}]
+  - elsif page_mode == "feature-starwars"
+    - heroes = [{:id=>2, :id_s=>"hero_starwars", :type_s=>"widehero", :param1_s=>"15% 50%", :param2_s=>"celeba", :param3_s=>"bottom", :image_t=>"/images/homepage/starwars.jpg", :site_s=>"both", :active_s=>"show", :language_supported_s=>"all"}]
```
Comment (breville) [inline pegasus/sites.v3/code.org/views/homepage_hero.haml]: Ideally we could go back to using gsheet-driven attributes now.

File: pegasus/sites.v3/code.org/views/learn_carousel_block.haml [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+- show_sw_blocks = id == "starwars" && (request.locale != "en-US" || mobile?)
+- if show_sw_blocks
+  - launch_code = "starwarsblocks"
+- else
+  - launch_code = id
+
+.container-fluid
+  .row-fluid
+    .courseblock-tall
+      %a{href: "http://studio.code.org/s/#{id}/reset"}
```
Comment (breville) [inline pegasus/sites.v3/code.org/views/learn_carousel_block.haml]: This should also use CDO.

File: pegasus/sites.v3/code.org/views/learn_carousels.haml [hunk 16]
```diff
@@ -32,7 +23,7 @@
     %button{:style=>"font-size:20px;"}
       = I18n.t(:carousel_heading_domestic)
 
-#all-tutorials{:style=>'visibility:hidden; overflow:hidden'}
+#all-tutorials{:style=>'visibility:hidden; overflow:hidden; margin-top: -30px;'}
```
Comment (breville) [inline pegasus/sites.v3/code.org/views/learn_carousels.haml:23]: is there a .css file where this margin adjustment should go?

File: pegasus/sites.v3/code.org/views/learn_code_studio.haml [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+.clear
+%br
+
+:javascript
+  $('#codeStudioCarousel').carousel({
+    interval: false
+  })
+
+  $('.multi-item-carousel .item').each(function() {
+    console.log("here")
```
Comment (breville) [inline pegasus/sites.v3/code.org/views/learn_code_studio.haml]: console.log breaks on IE9 if dev console isn't open, so we shouldn't have these here.  (or if we really, really need them, check that console/console.log actually exist.)

File: pegasus/sites.v3/code.org/views/learn_code_studio.haml [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+    var next = $(this).next();
+    if (!next.length) {
+      console.log("end of row")
+      next = $(this).siblings(':first');
+    }
+    next.children(':first-child').clone().appendTo($(this));
+
+    if (next.next().length > 0) {
+      console.log("if")
+      next.next().children(':first-child').clone().appendTo($(this));
```
Comment (breville) [inline pegasus/sites.v3/code.org/views/learn_code_studio.haml:91]: some comments would probably be useful here.

File: pegasus/sites.v3/code.org/public/mc.haml (hunk 3 above)
Comment (tanyaparker) [inline pegasus/sites.v3/code.org/public/mc.haml:26]: We only had it on the /mc page because we wanted it to be hidden until we flipped the switch. We can just show everything now, doesn't matter what the page mode is here.

File: pegasus/sites.v3/code.org/public/index.haml (hunk 2 above)
Comment (tanyaparker) [inline pegasus/sites.v3/code.org/public/index.haml:55]: I removed it from the pages where we only needed it before we flipped the switch. I also had to move it the homepage top level because when it was called in homepage_hero and homepage_sections they could be different. Then you would get the same tutorial in both the hero and hoc block.

File: pegasus/sites.v3/code.org/views/homepage_hero.haml (hunk 4 above)
Comment (tanyaparker) [inline pegasus/sites.v3/code.org/views/homepage_hero.haml]: Yes, the hero on homepage and banner on /learn will animate between the two. todo

**Comment by tanyaparker** [comment]
@caleybrock in reply to screenshots
- I just think it looks really weird to have the arrows between the title and block. I don't think it's as important to emphasize that.
- I got rid of the bee banner for smaller sizes, yeah, Hadi said it's not as important and can be get rid of for more vertical height. I also decreased the margin between the tabs and the bee, I thought it was cute how it overlapped like they're together as one thing not two separate things. It's just part of his wing and doesn't overlap the text.

File: pegasus/sites.v3/code.org/views/learn_carousels.haml (hunk 6 above)
Comment (tanyaparker) [inline pegasus/sites.v3/code.org/views/learn_carousels.haml:1]: No, that's Scratch, Tynker, and a lot of others. They're still important. We're removing the Code.org tutorials from this category.

**Comment by tanyaparker** [comment]
starwars_banner_mobile.jpg was just a temp, we're not using a banner on mobile anymore I'm going to delete.

File: pegasus/sites.v3/code.org/views/learn_code_studio.haml (hunk 17 above)
Comment (tanyaparker) [inline pegasus/sites.v3/code.org/views/learn_code_studio.haml]: My bad, was part of debug. Will remove as part of cleanup.

File: pegasus/sites.v3/code.org/public/hourofcode/artist.md (hunk 12 above)
Comment (tanyaparker) [inline pegasus/sites.v3/code.org/public/hourofcode/artist.md:37]: I'll work on updating Melissa's pages w/ video partial & CDO links in a separate PR.

File: pegasus/sites.v3/code.org/views/learn_carousel_block.haml (hunk 5 above)
Comment (tanyaparker) [inline pegasus/sites.v3/code.org/views/learn_carousel_block.haml:20]: Now tested and added overflow/ellipsis rules. Had to change the button to fit below the links instead. I think it still works.

File: pegasus/sites.v3/code.org/public/css/learn-carousel.css (hunk 10 above)
Comment (tanyaparker) [inline pegasus/sites.v3/code.org/public/css/learn-carousel.css:59]: Yes, works in IE8 now! Phew.

**Comment by tanyaparker** [comment]
@caleybrock @breville made code review changes -- pending these tests pass I'm going to merge into staging so you all can easily test and try out stuff.
